Show eventless contact_links and relation #3892

phebal opened the pull request:
Ready for review.

Commit:
Show eventless contact_links and relation #3892
Display ContactLinks that are not linked to any event in
Participant's 'Contact History' and contact_links index.
List Relationship to the Participant when listing
ContactLinks if other then self. #3596
Only display the ContactLinks and Events of the Participant
in their "Contact History' page and not those of other, related
participants.

hannahwhy commented on spec/controllers/contacts_controller_spec.rb:
```diff
@@ -218,6 +218,81 @@ def mock_event(stubs={})
+              "contact_language_code"=>"1",
+              "contact_interpret_code"=>"-3",
+              "contact_comment"=>""
+            },
```
Do you really need _all_ of these attributes?

hannahwhy commented:
Why is it necessary to munge contact links and events in the same array?  

You can get an event from a contact link.  If you want to display contact links for a participant regardless of whether it has an event (which appears to be the point here), the data retrieval problem is really just this:

``` ruby
ContactLink.includes(:event).where(:participant_id => @participant.id)
```

The view can then show event data based on whether or not the contact link has an event.

hannahwhy commented on app/assets/stylesheets/partials/_participants.scss:
```diff
@@ -24,7 +24,14 @@
+    margin-bottom: 0.75em;
+    padding: 0.25em;
+  }
+
```
This style is almost identical to the style for `.event`, except for extra padding.

Why do you need the duplication?  Can you eliminate this style?

rsutphin commented:
> Why is it necessary to munge contact links and events in the same array? 

It matches [the desired UI](https://code.bioinformatics.northwestern.edu/issues/issues/show/3892). It's not a table of contacts, some of which have events. It's a table of events which have contacts nested within them. This feature adds eventless contacts as peers of events.

rsutphin commented on app/views/people/_events_and_contacts.html.haml:
```diff
@@ -1,13 +1,18 @@
-        - participant.events.chronological.reverse_order.each do |event|
+  .participant_contacts
+    - if @events_and_contacts.blank?
+      = "No events or contacts exist for #{blank_safe(@person.to_s)} at this time."
```
Not your addition, but let's fix it anyway: "at this time" is unnecessary. "Exist" carries that implication.

rsutphin commented on app/views/people/_events_and_contacts.html.haml:
```diff
@@ -1,13 +1,18 @@
 .page_section
-  - @person.participants.each do |participant|
-    .participant_events
```
It looks like there's a behavior change here. Formerly, this list included all events for all the participants associated with the subject participant's person. That means that a child's events would show up on a mother's participant page and vice versa. Unless I missed something, `@events_and_contacts` only includes information for the subject participant — i.e., just the child or just the mother. This should be corrected. (Or I should be corrected.)

pfriedman replied:
This might address issue [3596](https://code.bioinformatics.northwestern.edu/issues/issues/show/3596) but I'm not certain if that was the intent of the change.

rsutphin replied:
It's probably also worth noting that the previous behavior created separate event history blocks for the participant and any related participants. It did not sort them together and, AFAICT, did not identify which participant each set was for. I don't know if those choices were due to user feedback or were just the first thing that got implemented.

rsutphin replied:
Addressing Cases-3596 along with this makes a lot of sense to me.

phebal replied:
Yes, this is meant to address both 3892 and 3596, I should have been more clear about this.  UI for 3596 is based on Rhett's and Paul's comments: https://code.bioinformatics.northwestern.edu/issues/issues/show/3596#note-4

Here's a screenshot of the UI for a participant:
![](http://i.imgur.com/FIUXUli.png?1)
And participant's child:
![](http://i.imgur.com/m0ZMFxA.png?1)

rsutphin replied:
Thanks, @phebal. That looks great. Can you add the relationship to the event also, using the relationship from the subject participant to the event's participant (and, as with the contacts, leaving it blank if the relationship is self)? E.g.:

Birth 2013-02-28 - ______ Biological Mother

hannahwhy commented on app/controllers/participants_controller.rb:
```diff
@@ -39,10 +39,38 @@ def in_ppg_group
+
+    rows.uniq.flatten.sort_by do |r|
+      find_date_and_time(r)
+    end
```
Sentinel-free code that can be adapted for other purposes:

``` ruby
class Event
  def started_at
    [event_start_date, event_start_time].map { |t| t || Unspecified }
  end
end

class ContactLink
  def started_at
    [contact_date_date, contact_start_time].map { |t| t || Unspecified }
  end
end

class Unspecified
  def self.<=>(other)
    1
  end
end

class ParticipantsController
  def get_sorted_events_contact_links
    objects = # ...
    objects.sort_by(&:started_at)
  end
end
```

The trick is finding an appropriate name and place for `Unspecified`.  That said, I'm pretty sure I've seen this "shove nils at the end of the list" elsewhere in Cases' views, which means it could live in a helper.

hannahwhy replied:
> which means it could live in a helper

That is:

``` ruby
class Event
  def started_at
    [event_start_date, event_start_time]
  end
end

class ParticipantsController
  def show
    @events_and_contacts = events_and_contact_links
  end

  def events_and_contact_links
    @participants.contact_links.collect # etc
  end
end

module ListHelper
  def sort_by_start(list)
    list.sort_by { |x| x.started_at.map { |dt| dt || Unspecified } }
  end

  class Unspecified
    def self.<=>(other); 1; end
  end
end
```

and in your view:

``` haml
- sort_by_start(@events_and_contacts).each do |o|
  # ...
```

phebal replied:
This is great and I restructured the code to look like this.  However using `Unspecifed` for sorting doesn't seem to work without also overloading `String` and `Date's` `<=>` operator.  I think it's less elegant but simpler to use a private method that returns a future date and time instead.

rsutphin commented on app/controllers/participants_controller.rb:
```diff
@@ -39,10 +39,38 @@ def in_ppg_group
+  def get_sorted_events_contact_links
+    rows = @participant.contact_links.collect(&:event).compact +
+           @participant.events +
+           @participant.person.contact_links.select {|cl| cl.event.nil?}
```
`@participant.contact_links` and `@participant.person.contact_links` are (unfortunately) the same collection. Could you use `@participant.person.contact_links` for both? I think that will make it clearer that the first and the third collections are complementary.

rsutphin commented on app/models/person.rb:
```diff
@@ -473,6 +473,18 @@ def first_child?
+    NcsCode.for_list_name_and_local_code(
+      'PERSON_PARTCPNT_RELTNSHP_CL1', rel_code
+    ).display_text
+  end
```
This could be a little simpler:

``` ruby
rel = person.participant_person_links.find { |ppl| ppl.participant == participant }.try(:relationship)

rel.try(:display_text) || 'None'
```

rsutphin replied:
(The simpler solution also has the benefit of using the model-configured code list. As you know, those can vary between MDES versions.)

phebal commented:
I believe my latest amend addresses all the comments and adds the relationship display for events.

New UI demonstration this:
![](http://i.imgur.com/5it5N0L.png?1)

rsutphin commented:
Looks good to me.
